### PR TITLE
Flutter 1.24, attempt 2

### PR DESCRIPTION
**Pull request by domesticmouse**

### CODE REVIEW

**Comment by domesticmouse** [comment]
Hey @RedBrogdon, @devoncarew and @nshahan,

This is a follow on from the conversation in #620, except this time I'm attempting to land Flutter 1.24 into a Dart Services environment with all the tests, as opposed to the original state where I was attempting to land flailing tests into an already landed Dart Services with Flutter 1.24. 

This time the failing tests show null safety issues clearly, even with the change to the generated `pubspec.yaml`.

Given I can't force `dartdevc` to not do null safety via command line args, and the `pubspec.yaml` we initialise the samples with doesn't seem to effect compilation, I suspect we are stuck on Flutter 1.23 until we decide to go full null safe on Dartpad.

**Comment by nshahan** [comment]
OK I gathered a few more details that could help resolve the issue here and forgive me if we already went over this before the holiday (all of that is a little foggy in my mind now). Are you running `pub get` in the directory where you generate that pubspec? 

That **_should_** force a package_config.json to be generated with the correct opt-out status in it. If it is still not working correctly could you print file `.dart_tool/package_config.json` somewhere in the tests before they fail so we can see the contents?

**Comment by domesticmouse** [comment]
@nshahan apologies for the delay in responding to this, I'm currently dealing with a situation with AppEngine Flex in that I can't deploy this app to GAE at all. I'll generate a `package_config.json for you soon.

**Comment by domesticmouse** [comment]
Picking one of the failing tests, the exec line looks something like:

```
/Users/brettmorgan/Documents/GitHub/dart-services/flutter/bin/cache/dart-sdk/bin/dartdevc --modules=amd \
  -s artifacts/flutter_web.dill \
  -s /Users/brettmorgan/Documents/GitHub/dart- 
services/flutter/bin/cache/flutter_web_sdk/flutter_web_sdk/kernel/flutter_ddc_sdk.dill \
  -o /var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpadPNabxu/main.dart.js --module-name dartpad_main \
  /var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpadPNabxu/bootstrap.dart \
  --packages=/var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpad6JamPl/.packages
```

Things to note, we have the main dart file, and a bootstrap wrapper, in a different directory to the .packages file. Here's the .packages file:

```
$ cat /var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpad6JamPl/.packages
# Generated by pub on 2020-12-07 15:44:59.867750.
async:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/async-2.5.0-nullsafety.3/lib/
boolean_selector:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/boolean_selector-2.1.0-nullsafety.3/lib/
characters:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/characters-1.1.0-nullsafety.5/lib/
charcode:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/charcode-1.2.0-nullsafety.3/lib/
clock:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/clock-1.1.0-nullsafety.3/lib/
collection:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/collection-1.15.0-nullsafety.5/lib/
fake_async:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/fake_async-1.2.0-nullsafety.3/lib/
flutter:file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/packages/flutter/lib/
flutter_test:file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/packages/flutter_test/lib/
matcher:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/matcher-0.12.10-nullsafety.3/lib/
meta:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/meta-1.3.0-nullsafety.6/lib/
path:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/path-1.8.0-nullsafety.3/lib/
sky_engine:file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/bin/cache/pkg/sky_engine/lib/
source_span:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/source_span-1.8.0-nullsafety.4/lib/
stack_trace:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/stack_trace-1.10.0-nullsafety.6/lib/
stream_channel:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/stream_channel-2.1.0-nullsafety.3/lib/
string_scanner:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/string_scanner-1.1.0-nullsafety.3/lib/
term_glyph:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/term_glyph-1.2.0-nullsafety.3/lib/
test_api:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/test_api-0.2.19-nullsafety.6/lib/
typed_data:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/typed_data-1.3.0-nullsafety.5/lib/
vector_math:file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/vector_math-2.1.0-nullsafety.5/lib/
dartpad_sample:lib/
```

Here is the `packages.json` file, but note it isn't from the same directory as the dart files:

```
$ cat /var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpad6JamPl/.dart_tool/package_config.json 
{
  "configVersion": 2,
  "packages": [
    {
      "name": "async",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/async-2.5.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "boolean_selector",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/boolean_selector-2.1.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "characters",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/characters-1.1.0-nullsafety.5",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "charcode",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/charcode-1.2.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "clock",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/clock-1.1.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "collection",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/collection-1.15.0-nullsafety.5",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "fake_async",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/fake_async-1.2.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "flutter",
      "rootUri": "file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/packages/flutter",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "flutter_test",
      "rootUri": "file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/packages/flutter_test",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "matcher",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/matcher-0.12.10-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "meta",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/meta-1.3.0-nullsafety.6",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "path",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/path-1.8.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "sky_engine",
      "rootUri": "file:///Users/brettmorgan/Documents/GitHub/dart-services/flutter/bin/cache/pkg/sky_engine",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "source_span",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/source_span-1.8.0-nullsafety.4",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "stack_trace",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/stack_trace-1.10.0-nullsafety.6",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "stream_channel",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/stream_channel-2.1.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "string_scanner",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/string_scanner-1.1.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "term_glyph",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/term_glyph-1.2.0-nullsafety.3",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "test_api",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/test_api-0.2.19-nullsafety.6",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "typed_data",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/typed_data-1.3.0-nullsafety.5",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "vector_math",
      "rootUri": "file:///Users/brettmorgan/.pub-cache/hosted/pub.dartlang.org/vector_math-2.1.0-nullsafety.5",
      "packageUri": "lib/",
      "languageVersion": "2.12"
    },
    {
      "name": "dartpad_sample",
      "rootUri": "../",
      "packageUri": "lib/",
      "languageVersion": "2.10"
    }
  ],
  "generated": "2020-12-07T04:44:59.881831Z",
  "generator": "pub",
  "generatorVersion": "2.12.0-29.10.beta"
}
```

And the `pubspec.yaml`:

```
$ cat /var/folders/6r/rk5z2g4521gfhwf4jw704pvh003nrc/T/dartpad6JamPl/pubspec.yaml 
name: dartpad_sample
environment:
  sdk: '>=2.10.0 <3.0.0'
dependencies:
  flutter:
    sdk: flutter
  flutter_test:
    sdk: flutter
```

**Comment by nshahan** [comment]
I suspect that if the `package_config.json` is in a different directory from the lib directory that contains `main.dart` and `bootstrap.dart` then it's not going to work as intended. Can you make sure it gets copied to the same temp directory? I think it's probably easiest to just copy it to the new temp directory that gets created for every run since it contains a relative path to the lib directory already.

Also, you should be able to pass it directly with the `--packages` flag instead of using a `.packages` file which is only supported for legacy support reasons.

**Comment by domesticmouse** [comment]
Thanks @nshahan that set of hints got me out of trouble. @RedBrogdon do you have time for a review?

File: lib/src/compiler.dart [hunk 1]
```diff
@@ -65,24 +65,28 @@ class Compiler {
     _logger.info('Temp directory created: ${temp.path}');
 
     try {
+      await copyPath(_flutterWebManager.dartTemplateProject.path, temp.path);
+
```
Comment (RedBrogdon) [inline lib/src/compiler.dart:69]: nit: whitespace

File: lib/src/compiler.dart [hunk 2]
```diff
@@ -128,9 +132,17 @@ class Compiler {
...
     try {
       final usingFlutter = _flutterWebManager.usesFlutterWeb(imports);
+      if (usingFlutter) {
+        await copyPath(
+            _flutterWebManager.flutterTemplateProject.path, temp.path);
+      } else {
+        await copyPath(_flutterWebManager.dartTemplateProject.path, temp.path);
+      }
+
+      await Directory(path.join(temp.path, 'lib')).create(recursive: true);
```
Comment (RedBrogdon) [inline lib/src/compiler.dart:142]: Also a nit: Consider creating `lib` as part of setting up the project directories in `grind.dart` and cutting this and similar lines.

File: lib/src/flutter_web.dart [hunk 3]
```diff
@@ -4,67 +4,21 @@
...
-    // and create a lib/ folder for completeness
-    Directory(path.join(_projectDirectory.path, 'lib')).createSync();
-  }
-
-  Future<void> warmup() async {
-    try {
-      if (_initedFlutterWeb) {
-        return;
-      }
+  final Directory dartTemplateProject = Directory(
```
Comment (RedBrogdon) [inline lib/src/flutter_web.dart:18]: Not strictly a result of this PR, but after it's landed, FlutterWebManager will have a weird set of responsibilities, not all of which actually relate to Flutter. This class could probably do with a refactor into several or at least a rename to something like "ProjectStructureManager" or something.

**Review by RedBrogdon — APPROVED** [review]
LGTM. My only real concern is the bit about FlutterWebManager now looking like an oddly patched-together class.

File: lib/src/compiler.dart (hunk 1 above)
Comment (domesticmouse) [reply lib/src/compiler.dart:69]: Done

File: lib/src/compiler.dart (hunk 2 above)
Comment (domesticmouse) [reply lib/src/compiler.dart:142]: I'm always reticent to create empty directories and then hope that copying tools do the right thing with the empty directories. Git, for instance, really doesn't. I've seen various other tooling in my DevOps career fail in similar ways in the face of empty directories, so making the empty directories just before putting files in them is way safer, both from a debugging and programmers intent perspective. It would be too easy for future maintenance to delete the empty directories from grind and then wonder why compilation is mystically failing. 

tl;dr: having two things at distance in a code base magically dependent on each other is bad, maintenance wise.

File: lib/src/flutter_web.dart (hunk 3 above)
Comment (domesticmouse) [reply lib/src/flutter_web.dart:18]: Once Flutter web goes stable my intent is to gut all the dual SDK code from the codebase, at which point a lot of this code can be smushed into other classes. Having only one SDK will enable a bunch of simplifications, including using Tony's slim dart docker image process along with a pre-compiled server to reduce startup time.

File: lib/src/compiler.dart (hunk 2 above)
Comment (RedBrogdon) [reply lib/src/compiler.dart:141]: Fair points.
